### PR TITLE
Fix VM Migrate complete email when To field is nil.

### PR DESCRIPTION
**Pull request by billfitzgerald0120**
Made change to not send email when the To field is nil.
Added log message that email will not be sent.

https://bugzilla.redhat.com/show_bug.cgi?id=1486461

### CODE REVIEW

**Comment by billfitzgerald0120** [comment]
@miq-bot add_label bug, fine/yes

**Comment by billfitzgerald0120** [comment]
@miq-bot assign @gmcculloug

**Comment by billfitzgerald0120** [comment]
@tinaafitz Please Review

**Comment by gmcculloug** [comment]
@tinaafitz Changes are easier to review by ignoring whitespace: https://github.com/ManageIQ/manageiq-content/pull/177/files?w=1

File: content/automate/ManageIQ/Infrastructure/VM/Migrate/Email.class/__methods__/vmmigratetask_complete.rb [hunk 1]
```diff
@@ -27,25 +27,30 @@
 $evm.log("info", "VM Owner: #{owner.inspect}")
 
 # to_email_address from owner.email then from model if nil
-to = owner.email || $evm.object['to_email_address']
+to = owner.email unless owner.nil?
```
Comment (gmcculloug) [inline content/automate/ManageIQ/Infrastructure/VM/Migrate/Email.class/__methods__/vmmigratetask_complete.rb]: @billfitzgerald0120 Keep the original format and use `try` to handle the `nil` owner object.

`to = owner.try(:email) || $evm.object['to_email_address']`

**Comment by billfitzgerald0120** [comment]
@gmcculloug made changes as requested

File: spec/automation/unit/method_validation/vm_migrate_task_complete_spec.rb [hunk 2]
```diff
@@ -11,11 +11,21 @@
     expect(GenericMailer).to receive(:deliver).with(:automation_notification,
                                                     hash_including(:to   => user.email,
                                                                    :from => "evmadmin@example.com"
-                                                                  )
+                                                                )
```
Comment (gmcculloug) [inline spec/automation/unit/method_validation/vm_migrate_task_complete_spec.rb]: @billfitzgerald0120 The change in whitespace here is revealing the rubocop warnings around alignment.  Please update the closing parentheses, like you did in the new test below, to remove the warnings.

File: spec/automation/unit/method_validation/vm_migrate_task_complete_spec.rb [hunk 3]
```diff
@@ -11,11 +11,21 @@
...
                                                    )
     attrs = ["MiqServer::miq_server=#{miq_server.id}"]
     attrs << "MiqRequestTask::vm_migrate_task=#{miq_request_task.id}"
     attrs << "vm_migrate_task_id=#{miq_request_task.id}"
     MiqAeEngine.instantiate("/Infrastructure/VM/Migrate/Email/VmMigrateTask_Complete?event=vm_migrated&#{attrs.join('&')}", user)
   end
+
+  it 'sends email to nil' do
```
Comment (gmcculloug) [inline spec/automation/unit/method_validation/vm_migrate_task_complete_spec.rb]: Test name is misleading as you are expecting the logic to **not** send an email in this case.  Please reword.

**Comment by billfitzgerald0120** [comment]
@gmcculloug Made changes as requested, please review

**Comment by gmcculloug** [comment]
@billfitzgerald0120 @tinaafitz Looks like we have this exposure in a number of email methods.  Do we handle this anywhere else?  Would it make more sense to send the email to someone like an administrator in this case instead of just skipping the email when someone migrates a VM?

**Comment by billfitzgerald0120** [comment]
@gmcculloug @tinaafitz We handle this in Infrastructure and Cloud MiqProvision_Complete method.
Sending it to an administrator sounds better that skipping.

**Comment by billfitzgerald0120** [comment]
@tinaafitz I added the requester_email as the second choice for the 'to' email value.

**Comment by billfitzgerald0120** [comment]
@tinaafitz Please Review

**Comment by gmcculloug** [comment]
@tinaafitz Does the method need to change for this or can this be fixed in the schema?  I recall we briefly discussed this.

@billfitzgerald0120 Can you describe the condition in which this happens since the schema includes a default email address?  When would the `to` email address be empty?

**Comment by billfitzgerald0120** [comment]
@gmcculloug @tinaafitz The problem was the owner.email was failing.  I added the .try which fixed the problem.   I added the requester.email as the second option.  We talked about making sure that we send an email to someone and not just abort the email.  The schema should always have a value but the requester is probably a better option.  It's possible the schema has the default values for to_email_address instead of the actual requester.

-to = owner.email || $evm.object['to_email_address']
+to = owner.try(:email) || requester.email || $evm.object['to_email_address']

**Comment by billfitzgerald0120** [comment]
@tinaafitz Please review

**Comment by simaishi** [comment]
Fine backport details:

```
$ git log -1
commit a78b743cd0f2f857bace5192141d2f8b5e8864bd
Author: Greg McCullough <gmccullo@redhat.com>
Date:   Wed Sep 20 10:39:16 2017 -0400

    Merge pull request #177 from billfitzgerald0120/vmmigrate_email_fix
    
    Fix VM Migrate complete email when To field is nil.
    (cherry picked from commit 07fe288e98886d1c9bb98cf392a876c59b287413)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1496937
```

**Pull request merged**